cmd/admin-handlers.go

```diff
@@ -1303,6 +1304,154 @@ func (a adminAPIHandlers) KMSKeyStatusHandler(w http.ResponseWriter, r *http.Req
+	// if no bucket is specified list all buckets.
+	var bucketNames []string
+	if req.Bucket != "" {
+		if _, err := objectAPI.GetBucketInfo(ctx, req.Bucket); err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = []string{req.Bucket}
+	} else {
+		bucketInfo, err := objectAPI.ListBuckets(ctx)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = make([]string, len(bucketInfo))
+		for i, b := range bucketInfo {
+			bucketNames[i] = b.Name
+		}
+	}
+
+	// For each bucket list all objects for the requested prefix.
+	// If the object is encrypted with the KMS perform a key
+	// rotation if there is a mapping for the current key ID.
+	for _, bucket := range bucketNames {
+		var err error
+		var objectListing ListObjectsV2Info
+		for {
+			objectListing, err = objectAPI.ListObjectsV2(ctx, bucket, req.Prefix, objectListing.NextContinuationToken, "", 1000, false, objectListing.ContinuationToken)
+			if err != nil {
+				writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+				return
+			}
+
+			for _, object := range objectListing.Objects {
+				if !req.Recursive && object.IsDir {
+					continue
+				}
+
+				if crypto.IsEncrypted(object.UserDefined) && crypto.S3.IsEncrypted(object.UserDefined) {
+					keyID := object.UserDefined[crypto.S3KMSKeyID]
+					newKeyID, ok := req.KeyMapping[keyID]
+					if !ok {
+						continue // No key mapping
+					}
+
+					objectKey, err := crypto.S3.UnsealObjectKey(GlobalKMS, object.UserDefined, bucket, object.Name)
+					if err != nil {
+						writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+						return
+					}
+					newKey, newEncKey, err := GlobalKMS.GenerateKey(newKeyID, crypto.Context{bucket: path.Join(bucket, object.Name)})
+					if err != nil {
+						writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+						return
+					}
+					sealedKey := objectKey.Seal(newKey, crypto.GenerateIV(rand.Reader), crypto.S3.String(), bucket, object.Name)
+					object.UserDefined = crypto.S3.CreateMetadata(object.UserDefined, newKeyID, newEncKey, sealedKey)
+					object.metadataOnly = true
+
+					if !req.DryRun { // actually update the metadata on the backend
+						object, err = objectAPI.CopyObject(ctx, bucket, object.Name, bucket, object.Name, object, ObjectOptions{}, ObjectOptions{})
+						if err != nil {
+							writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+							return
+						}
+					}
+
+					rotateResponse := madmin.KMSRotateKeyResponse{
+						Bucket:   bucket,
+						Object:   object.Name,
+						OldKeyID: keyID,
+						NewKeyID: object.UserDefined[crypto.S3KMSKeyID],
+					}
+					if _, err = fmt.Fprintln(w, rotateResponse.JSON()); err != nil {
+						writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+						return
+					}
+				}
+			}
+
+			if !objectListing.IsTruncated {
+				break
+			}
+		}
```

Comment:
That is correct. I'm not sure I understand your concern. Returning e.g. only the first object that has been rotated is problematic. There can be multiple objects encrypted with different master keys. For example:
```
bucket1 |
              |- object1a <=> master-key-1
              |- object1b <=> master-key-2
              |- object1c <=> master-key-1

bucket2 |
              |- object2a <=> master-key-1
              |- object2b <=> master-key-2
              |- object2c <=> master-key-3
```
Now I may want to rotate all objects from `master-key-2` to `master-key-1` and `master-key-3` to master-key-4`. However, for some reason MinIO does not have access to `master-key-4` at the moment.

How would I know which objects have been rotated and that one rotation failed?

Reply:
@aead, I would expect the API to acknowledge each object that was key rotated and print the old and new key for that object in the response. The example program prints just the first object that was key rotated - pls fix the example. 

Aside of that fact, this API will time out with current style of implementation - for example all the objects in the store could have been encrypted with `oldkey` and you are attempting to rotate recursively.

Reply:
The API does that. The example code (as you mentioned) was wrong. Fixed.

Rotating all objects should be a valid operation. Why do you think it would timeout? The client gets a response stream.